### PR TITLE
WIP: Support layers for managing many workspaces

### PR DESCRIPTION
**Pull request by bastidest**
## Proposal
I want to manage large numbers of workspaces (>20), which can be grouped into different layers (default, work, uni, ...) as they are completely different tasks. I rarely switch **between layers**, but I switch **inside layers** very frequently.

Every layer has a similar structure: WS1 = webbrowser, WS2 = emacs, WS3 = docs, ...

One workspace shall be reachable from every layer, as it contains email, password manager, ...

The idea is to define layers, which make switching workspaces inside them easier. 

First of all: The original behavior obviously needs to stay the same. If one chooses to define a layer, it needs to be added to the configuration. A layer is defined with the directive `layer <name> <from> <to>`.

## Example
output
```
# switch to workspace
bindsym $mod+1 workspace number 1
bindsym $mod+2 workspace number 2
# ...
# layer configuration
layer uni 11 19
layer work 21 29
bindsym $mod+F1 layer default
bindsym $mod+F2 layer uni
bindsym $mod+F3 layer work
# ...
```
This defines a layer "uni" starting at workspace 11, ending at workspace 19 (inclusive) and a layer "work" starting at 21, ending at 29. Also there are keybindings for switching to the defined layers.

### Switching
- At startup, layer "default" is enabled, wich has no influence on any behavior
- Pressing `$mod+F1` will activate layer "uni", but do nothing else
- Upon pressing `$mod+1`, the relative offset of the layer will be added to the switching command: `workspace number 1` -> `workspace number 11`
- Upon pressing `$mod+2`, the relative offset of the layer will be added to the switching command: `workspace number 2` -> `workspace number 12`
- ...
- Upon pressing `$mod+9`, the relative offset of the layer will be added to the switching command: `workspace number 9` -> `workspace number 19`
- Upon pressing `$mod+0`, the switch will happen normally, as the last mapped workspace is `19`

### Moving
"I want to move the current container to the 2nd workspace of layout 'work'" == "I want to move the current container to workspace 22"
- `$mod+F3` == `layer work`
- `$mod+shift+2` == `move container to workspace number 2`

## TODOs
- [ ] See if there is any form of acceptance here, or if I need to maintain this seperately
- [ ] Tests
- [ ] Concept Improvements

### CODE REVIEW

**Comment by rayvermey** [comment]
This is really a very nice add on for i3! 
Would be nice if this can be incorporated in the main i3!
Ray

**Comment by MatheusMuriel** [comment]
Wow its very nice!

**Comment by NightH4nter** [comment]
This would be a pretty good addition. Ideally if it's gonna be compatible with i3-gaps.

**Comment by existme** [comment]
Great addition, thank you. Key binding for many workspaces opens up i3 for lots of new possibilities

**Comment by matthinc** [comment]
Good idea! I really hope that this will become an official feature!

**Comment by NightH4nter** [comment]
Probably another TODO is to make it possible to move a window from a workspace on one layer to a workpace on another layer.

**Comment by bastidest** [comment]
@NightH4nter That already works. To move a window to another layer, just switch to the layer and press your "move workspace" command.

Example: "I want to move the current container to workspace 23" -> $mod+FN2, $mod+shift+3

**Comment by NightH4nter** [comment]
> @NightH4nter That already works. To move a window to another layer, just switch to the layer and press your "move workspace" command.
> 
> Example: "I want to move the current container to workspace 23" -> $mod+FN2, $mod+shift+3

Okay, I understand. Nice!

**Comment by jpsala** [comment]
I like it!

**Comment by Airblader** [comment]
Thanks for bringing this up and making a PR! We have seen questions surrounding this topic a few times in the past and I can see the value and am open to such a concept. However, I'm not really happy with the specific proposal in this PR as it uses a very static configuration, and I can already see that people would want to modify this at runtime. 

Out of curiosity: what happens if I have layers 1—10, 11—20, 21—30 and while in the first layer I open workspace 15? Simply using an offset will overlap this into another layer, won't it?

A feature like this comes with quite a bit of configuration and/or command surface, which means it is a significant complexity driver. Therefore I think we need to carefully discuss an approach for this. 

For example, we could omit any kind of upfront config and instead just have a `layer foo` command which creates the layer on the fly. All workspace management happens entirely inside the layer, no matter the workspace number etc. Then there is a single `move workspace to layer` command. Immediate questions coming to my mind:

1. Do we even really need to support moving windows between layers individually? This would be a significant driver in the command surface. Using command criteria like marks, con_id etc would allow this without modification on the command syntax.
2. How to handle a new window which relates to layer A while we are on layer B? Do we show it in layer A (= not visible to user), layer B (= breaks layer assignment for the app), automatically switch to layer A and show it there (= can be annoying)? Does it have to be configurable?
3. Do we treat window activations on other layers just like we currently do window activations on non visible workspaces?
4. How to handle specific requests to focus a window on another layer? Automatically switch to it? Or reject the request? I think this one is tricky, the correct answer probably is "it depends".
5. On EWMH/ICCCM side we *must* expose all workspaces regardless of group I think or else we'd break specs. This will cause bars to always show all workspaces unless they implement specific support for it (unfortunately I don't think there is a spec for this to implement against). i3bar would of course be such a client that needs to handle it. 
6. Given that, can something like stickiness even be kept in a spec-compliant way? I assume we don't want sticky across layers (would also be even more complex), but not doing so probably breaks its spec as the window will not be shown on its output. We have to read the spec to see if this is OK, or discuss not allowing sticky on inactive layers.
7. Do we keep workspace to output assignments as-is? This means this config is valid across all layers individually.

There's probably a whole lot more.

---

In summary I hope you can see this isn't too trivial of an issue. However, I do hear the support for it in the community. Therefore I will close this PR and ask for someone to open a feature request issue with a proposal and ideally addressing various scenarios with examples. A topic like this needs to he spec'd out in a lot of detail before we can ship it. At this stage the PR is not yet helpful I think as we first need to find a good direction. Thanks!

**Comment by orestisfl** [comment]
I believe this functionality can be implemented through an IPC script, which would be more flexible and maintainable.

> This will cause bars to always show all workspaces unless they implement specific support for it

For this, we have https://github.com/i3/i3/issues/2333 / https://github.com/i3/i3/issues/3818

**Comment by Airblader** [comment]
> I believe this functionality can be implemented through an IPC script

This was my initial thought also, but I'm actually not sure how convenient that would be. But definitely worth a shot and in fact this has been done before (I never tried it out myself though).

**Comment by bastidest** [comment]
@Airblader 

Thank you for taking the time to read the proposal! I too was surprised by the community feedback.

> Out of curiosity: what happens if I have layers 1—10, 11—20, 21—30 and while in the first layer I open workspace 15? Simply using an offset will overlap this into another layer, won't it?

As there is only one active layer at a time, the concept of "other layers" does not exist. When layer one (1-10) is enabled and you switch to WS15, you switch to WS15. The layer only affects commands in the range.

> A feature like this comes with quite a bit of configuration and/or command surface, which means it is a significant complexity driver. Therefore I think we need to carefully discuss an approach for this.

I completely agree on that, it was meant as a immediate solution for me and as input for the community.

> Then there is a single `move workspace to layer` command.

This is the exact opposite of my idea, as the layers are completely transparent. Adding another non-concept would require too much modifications. For more elaborate concepts there is always IPC.

> Do we even really need to support moving windows between layers individually?

See above.

> How to handle a new window which relates to layer A while we are on layer B?

Again. No handling needed, as the layers are a transparent concept.

> Do we treat window activations on other layers just like we currently do window activations on non visible workspaces?

Exactly. The only thing the layers do is to give a visual indication of the current layer range and make switching inside a layer easier.

> How to handle specific requests to focus a window on another layer?

Valid point. From my perspective, I'd never touch the current active layer when doing anything.

I think you get my point. My idea was to keep the layers as non-invasive as possible in order to not change any existing behavior.

@orestisfl 
> I believe this functionality can be implemented through an IPC script, which would be more flexible and maintainable.

In fact this was my first shot at the problem, but I quickly hit obstacles. Im sure it can be done, but having to rely on a external daemon for correct window switching behavior is not acceptable to me. Also I don't know how far the i3bar can be customized to show the current active layer with their respective workspaces.

Thank you everyone for taking the time. I will definetly keep the fork updated, as I now depend on it 😅

**Comment by Ricocotam** [comment]
Do you plan opening an issue so we can all progress on this awesome feature ?

**Comment by bastidest** [comment]
> Do you plan opening an issue so we can all progress on this awesome feature ?

Will do in this week.

**Comment by bastidest** [comment]
I created an issue for further discussing and clarifying this idea. Leave your thoughts on that topic there. #4024

**Comment by rayvermey** [comment]
Thanks, I will.

Ray

Verzonden vanuit Outlook Mobile<https://aka.ms/blhgte>

________________________________
